### PR TITLE
Add support for Gitlab groups and subgroups

### PR DESCRIPTION
**Pull request by miveronese**
#34922

This PR adds support to restrict user login based on a list of  allowed GitLab groups or subgroups that can be informed in the configuration for this auth provider.

Here's how the new field, added under GitLabAuthProvider, looks like:

```json
   "allowGroups": {
          "description": "...",
          "default": [],
          "type": "array",
          "items": {
            "type": "string",
            "minLength": 1
          },
          "examples":[["group", "group/subgroup", "group/subgroup/subgroup"]]
        }
```

## Test Plan

New unit tests were added  and some existing tests were updated to cover the changes in this PR.

### CODE REVIEW

File: enterprise/cmd/frontend/internal/auth/oauth/middleware.go [hunk 1]
```diff
@@ -72,7 +73,8 @@ func newOAuthFlowHandler(db database.DB, serviceType string) http.Handler {
 		p := GetProvider(serviceType, id)
 		if p == nil {
 			log15.Error("no OAuth provider found with ID and service type", "id", id, "serviceType", serviceType)
-			http.Error(w, "Misconfigured GitHub auth provider.", http.StatusInternalServerError)
+			msg := fmt.Sprintf("Misconfigured %s auth provider.", serviceType)
+			http.Error(w, msg, http.StatusInternalServerError)
```
Comment (miveronese) [inline enterprise/cmd/frontend/internal/auth/oauth/middleware.go:77]: Updated this because there's a misleading error message.

File: doc/admin/auth/index.md [hunk 2]
```diff
@@ -134,6 +134,7 @@ Then add the following lines to your site configuration:
       "clientSecret": "replace-with-the-oauth-client-secret",
       "allowSignup": false,  // CAUTION: Set to true to enable signup. If nothing is specified in `allowOrgs`, any GitHub user can sign up.
       "allowOrgs": ["your-org-name"] // Restrict logins and signups if enabled to members of these orgs.
```
Comment (unknwon) [inline doc/admin/auth/index.md:136]:
```suggestion
      "allowOrgs": ["your-org-name"], // Restrict logins and signups if enabled to members of these orgs.
```

File: doc/admin/auth/index.md [hunk 3]
```diff
@@ -175,7 +176,8 @@ Then add the following lines to your site configuration:
         "displayName": "GitLab",
         "clientID": "replace-with-the-oauth-application-id",
         "clientSecret": "replace-with-the-oauth-secret",
-        "url": "https://gitlab.example.com"
+        "url": "https://gitlab.example.com",
+        "allowGroups": ["yourgroup-full-path", "your-group/yoursubgroup-full-path"] // Restrict logins to members of groups or subgroups based on the full-path provided.
```
Comment (unknwon) [inline doc/admin/auth/index.md:180]: The presence of "-full-path" feels confusing... I prefer the example you give in the PR description.

File: enterprise/cmd/frontend/internal/auth/gitlaboauth/session.go [hunk 4]
```diff
@@ -182,3 +194,35 @@ func (s *sessionIssuerHelper) SessionData(token *oauth2.Token) oauth.SessionData
 		// TODO(beyang): store and use refresh token to auto-refresh sessions
 	}
 }
+
+// verifyUserGroups checks whether the authenticated user belongs to one of the GitLab groups when the allowGroups option is set
```
Comment (unknwon) [inline enterprise/cmd/frontend/internal/auth/gitlaboauth/session.go:198]:
```suggestion
// verifyUserGroups checks whether the authenticated user belongs to one of the GitLab groups when the allowGroups option is set.
```

File: enterprise/cmd/frontend/internal/auth/gitlaboauth/session.go [hunk 5]
```diff
@@ -182,3 +194,35 @@ func (s *sessionIssuerHelper) SessionData(token *oauth2.Token) oauth.SessionData
...
+	if len(s.allowGroups) == 0 {
+		return true
+	}
+
+	allowed := make(map[string]bool, len(s.allowGroups))
+	for _, group := range s.allowGroups {
+		allowed[group] = true
+	}
+
+	for page := 1; hasNextPage; page++ {
```
Comment (unknwon) [inline enterprise/cmd/frontend/internal/auth/gitlaboauth/session.go:213]: nit:

```suggestion
	if len(s.allowGroups) == 0 {
		return true
	}

	allowed := make(map[string]bool, len(s.allowGroups))
	for _, group := range s.allowGroups {
		allowed[group] = true
	}

	var err error
	var gitlabGroups []*gitlab.Group
	hasNextPage := true
	for page := 1; hasNextPage; page++ {
```

File: enterprise/cmd/frontend/internal/auth/gitlaboauth/session.go [hunk 6]
```diff
@@ -182,3 +194,35 @@ func (s *sessionIssuerHelper) SessionData(token *oauth2.Token) oauth.SessionData
...
+		gitlabGroups, hasNextPage, err = glClient.ListGroups(ctx, page)
+		if err != nil {
+			return false
```
Comment (unknwon) [inline enterprise/cmd/frontend/internal/auth/gitlaboauth/session.go:216]: I think we need to print an error log before the `return`, otherwise the error is swallowed and going to be hard to debug. Or we should just make this method be able to return an error and propagate back to the client.

File: enterprise/cmd/frontend/internal/auth/gitlaboauth/session_test.go [hunk 7]
```diff
@@ -3,26 +3,198 @@ package gitlaboauth
...
+				description: "glUser, in allowed groups -> session created",
+				allowGroups: []string{"group1"},
+				glUser: &gitlab.User{
+					ID:       int32(101),
+					Username: string("alice"),
+					Email:    string("alice@example.com"),
+				},
+				glUserGroups: []*gitlab.Group{
+					{FullPath: "group1"},
+					{FullPath: "group2"},
```
Comment (unknwon) [inline enterprise/cmd/frontend/internal/auth/gitlaboauth/session_test.go:110]: Missing a case for subgroup?

File: enterprise/cmd/frontend/internal/auth/gitlaboauth/session_test.go [hunk 8]
```diff
@@ -3,26 +3,198 @@ package gitlaboauth
...
+
+					if uid, ok := authSaveableUsers[op.UserProps.Username]; ok {
+						return uid, "", nil
+					}
+
+					return 0, "safeErr", errors.New("error mocking get and save user")
+				}
+
+				defer func() {
+					auth.MockGetAndSaveUser = nil
```
Comment (unknwon) [inline enterprise/cmd/frontend/internal/auth/gitlaboauth/session_test.go:149]: Missing mock reset for `gitlab.MockListGroups1`.

File: enterprise/cmd/frontend/internal/auth/gitlaboauth/session_test.go [hunk 9]
```diff
@@ -3,26 +3,198 @@ package gitlaboauth
...
+
+func u(username, email string, emailIsVerified bool) database.NewUser {
+	return database.NewUser{
+		Username:        username,
+		Email:           email,
+		EmailIsVerified: emailIsVerified,
+	}
+}
+
+func acct(serviceType, serviceID, clientID, accountID string) extsvc.AccountSpec {
```
Comment (unknwon) [inline enterprise/cmd/frontend/internal/auth/gitlaboauth/session_test.go:189]: I think we are not getting much benefit for having `u` vs `newUser`, besides losing quite a lot readability.

```suggestion
func newUser(username, email string, emailIsVerified bool) database.NewUser {
	return database.NewUser{
		Username:        username,
		Email:           email,
		EmailIsVerified: emailIsVerified,
	}
}

func newAccountSpec(serviceType, serviceID, clientID, accountID string) extsvc.AccountSpec {
```

Comment (unknwon) [reply enterprise/cmd/frontend/internal/auth/gitlaboauth/session_test.go:189]: On second thought... I personally don't think these two functions have a reason to exist at all, they are literally composing structs without any extra logic.

File: internal/extsvc/gitlab/groups.go [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+	ID       int32  `json:"id"`
+	FullPath string `json:"full_path"`
+}
+
+var MockListGroups func(ctx context.Context, page int) ([]*Group, bool, error)
+
+// ListGroups returns a list of groups for the authenticated user.
+func (c *Client) ListGroups(ctx context.Context, page int) (groups []*Group, hasNextPage bool, err error) {
+	if MockListGroups != nil {
+		return MockListGroups(ctx, 1)
```
Comment (unknwon) [inline internal/extsvc/gitlab/groups.go:19]:
```suggestion
		return MockListGroups(ctx, page)
```

File: internal/extsvc/gitlab/groups_test.go [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+	t.Run("malformed response", func(t *testing.T) {
+		client := newTestClient(t)
+		client.httpClient = &mockHTTPResponseBody{
+			responseBody: `this is not valid JSON`,
+		}
+
+		groupsResponse, _, err := client.ListGroups(ctx, 1)
+		if groupsResponse != nil {
+			t.Error("unexpected non nil response")
```
Comment (unknwon) [inline internal/extsvc/gitlab/groups_test.go:47]:
```suggestion
			t.Error("unexpected non-nil response")
```

The other occurrence has the hyphen.

**Review by unknwon — APPROVED** [review]
I think the code works, but left a few comments, approve to unblock.

File: internal/extsvc/gitlab/groups.go (hunk 10 above)
Comment (miveronese) [reply internal/extsvc/gitlab/groups.go:19]: Hum, I thought I had pushed this change... good chatch!

File: enterprise/cmd/frontend/internal/auth/gitlaboauth/session.go (hunk 6 above)
Comment (miveronese) [reply enterprise/cmd/frontend/internal/auth/gitlaboauth/session.go:216]: Good point!

File: enterprise/cmd/frontend/internal/auth/gitlaboauth/session_test.go (hunk 7 above)
Comment (miveronese) [reply enterprise/cmd/frontend/internal/auth/gitlaboauth/session_test.go:122]: Added! 👍

**Comment by miveronese** [comment]
@unknwon , thanks for the review! I've updated this PR based on your suggestions.

**Review by unknwon — APPROVED** [review]
Thanks for the followup! Let's ship it.

File: doc/admin/auth/index.md [hunk 12]
```diff
@@ -175,7 +176,8 @@ Then add the following lines to your site configuration:
         "displayName": "GitLab",
         "clientID": "replace-with-the-oauth-application-id",
         "clientSecret": "replace-with-the-oauth-secret",
-        "url": "https://gitlab.example.com"
+        "url": "https://gitlab.example.com",
+        "examples":[["group", "group/subgroup", "group/subgroup/subgroup"]], // Restrict logins to members of groups or subgroups based on the full-path provided.
```
Comment (kopancek) [inline doc/admin/auth/index.md:180]: Should we call this `examples`? By looking at the config schema it seems like this should instead say `allowGroups`.

```suggestion
        "allowGroups":[["group", "group/subgroup", "group/subgroup/subgroup"]], // Restrict logins to members of groups or subgroups based on the full-path provided.
```

File: schema/site.schema.json [hunk 13]
```diff
@@ -1856,6 +1856,16 @@
           "description": "The OAuth API scope that should be used",
           "default": "api",
           "enum": ["api", "read_api"]
+        },
+        "allowGroups": {
```
Comment (kopancek) [inline schema/site.schema.json:1860]: I see this PR only adds `allowGroups` field. Do we have another PR that adds `allowSignUp` field?

**Review by kopancek — APPROVED** [review]
Added one suggestion for the docs page and one question. Otherwise looks OK to me.

File: schema/site.schema.json (hunk 13 above)
Comment (miveronese) [reply schema/site.schema.json:1860]: Yes, we have another [one](https://github.com/sourcegraph/sourcegraph/pull/36057).
